DirectEmployers/MyJobs
# [PD-1837] single page app - roles APIs

Please ignore front-end work at /manage-users/. That work will be completed in another PR.

This PR has 6 API endpoints that are relevant for: https://directemployersfoundation.atlassian.net/browse/PD-1837

One is for listing activities:
* manage-users/api/activities/

And five are for CRUDing of roles:
* manage-users/api/roles/create/
* manage-users/api/roles/
* manage-users/api/roles/NUMBER
* manage-users/api/roles/edit/NUMBER
* manage-users/api/roles/delete/NUMBER

To do:
* create automated tests for these end points?
* connect these end points up into the front end

[BTW, this Chrome extension is helpful for working with JSON: https://chrome.google.com/webstore/detail/jsonview/chklaanhfefbnpoihckbnefhakgolnmc?hl=en ]


To test each end point:

1) manage-users/api/activities/

Visit http://127.0.0.1:8000/manage-users/api/activities/

2) manage-users/api/roles/create/

Vist MyJobs local install: http://127.0.0.1:8000/
Open Chrome Dev Tools > Console

Copy/paste:

```javascript
// jQuery used for this example only
var jq = document.createElement('script');
jq.src = "//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js";
document.getElementsByTagName('head')[0].appendChild(jq);
var csrf = read_cookie("csrftoken");
```
And then:
```javascript
activities_array = ["delete user", "update user"]
users_array = ["dpoynter@apps.directemployers.org"]

$.post( "http://127.0.0.1:8000/manage-users/api/roles/create/",
  {
    csrfmiddlewaretoken: csrf, 
    role_name: "New Role",
    assigned_activities: activities_array,
    users: users_array
  },
  function( data ) {
    console.log( data );
});
```

Check that new role is created in myjobs_role table.

See that User has been associated with new role:
```python
User.objects.get(email="dpoynter@apps.directemployers.org").roles.all()
```
3) manage-users/api/roles/
Visit: http://127.0.0.1:8000/manage-users/api/roles/

4) manage-users/api/roles/NUMBER
Visit: http://127.0.0.1:8000/manage-users/api/roles/NUMBER

5) manage-users/api/roles/edit/NUMBER

Vist MyJobs local install: http://127.0.0.1:8000/

Copy/paste the following in Chrome Dev Tools > Console:
```javascript
// jQuery used for this example only
var jq = document.createElement('script');
jq.src = "//ajax.googleapis.com/ajax/libs/jquery/2.1.4/jquery.min.js";
document.getElementsByTagName('head')[0].appendChild(jq);
var csrf = read_cookie("csrftoken");
```
And then:
```javascript
activities_array = ["delete user", "update user"]
users_array = ["dpoynter@apps.directemployers.org"]

$.post( "http://127.0.0.1:8000/manage-users/api/roles/edit/8/",
  {
    csrfmiddlewaretoken: csrf, 
    role_name: "New Role",
    assigned_activities: activities_array,
    users: users_array
  },
  function( data ) {
    console.log( data );
});
```
Then confirm role name changed with something like:
```python
Role.objects.get(pk=8)
```
Then confirm activities altered with something like:
```python
Role.objects.get(pk=8).activities.all()
```
6) manage-users/api/roles/delete/NUMBER

Delete one of your new roles:
```javascript
$.ajax( "http://127.0.0.1:8000/manage-users/api/roles/delete/13/",
  {
    type: "DELETE",
    beforeSend: function(xhr) {
        xhr.setRequestHeader("X-CSRFToken", csrf);
    },
 success: function(data) { 
    console.log( data );
}});
```
